src/XrdTpc/XrdTpcNSSSupport.cc

```diff
@@ -0,0 +1,262 @@
+    const int m_output_fd;
+
+    decltype(&XrdCryptosslX509ToFile) m_x509_to_file_func;
+    int (*m_file_to_x509_func)(FILE *file, XrdCryptoX509Chain *c);
+};
+
+
+bool
+CASet::processFile(file_smart_ptr &fp, const std::string &fname)
+{
+    XrdCryptoX509Chain chain;
+    // Not checking return value here; function returns `0` on error and
+    // if no certificate is found.
+    (*m_file_to_x509_func)(fp.get(), &chain);
+
+    auto ca = chain.Begin();
+    // Note we purposely leak the outputfp here; we are just borrowing the handle.
+    FILE *outputfp = fdopen(m_output_fd, "w");
+    if (!outputfp) {
+        m_log.Emsg("CAset", "Failed to reopen file for output", fname.c_str());
+        return false;
+    }
+    while (ca) {
+        auto hash_ptr = ca->SubjectHash();
+        if (!hash_ptr) {
+            continue;
+        }
+        auto iter = m_known_cas.find(hash_ptr);
+        if (iter != m_known_cas.end()) {
+            //m_log.Emsg("CAset", "Skipping known CA with hash", fname.c_str(), hash_ptr);
+            ca = chain.Next();
+            continue;
+        }
+        //m_log.Emsg("CAset", "New CA with hash", fname.c_str(), hash_ptr);
+        m_known_cas.insert(hash_ptr);
+
+        if ((*m_x509_to_file_func)(ca, outputfp)) {
+            m_log.Emsg("CAset", "Failed to write out CA", fname.c_str());
+            return false;
+        }
+        ca = chain.Next();
+    }
+    fflush(outputfp);
+
+    return true;
+}
+
+}
+
+
+using namespace TPC;
+
+
+std::unique_ptr<XrdTpcNSSSupport::TempCAGuard>
+XrdTpcNSSSupport::TempCAGuard::create(XrdSysError &err) {
+    char fname[] = "/tmp/xrootd_curl_nss_ca.XXXXXX.pem";
+    int fd = mkstemps(fname, 4);
+    if (fd < 0) {
+        err.Emsg("NSSSupport", "Failed to create temp file:", strerror(errno));
+        return std::unique_ptr<TempCAGuard>();
+    }
+    return std::unique_ptr<TempCAGuard>(new TempCAGuard(fd, fname));
+}
+
+
+XrdTpcNSSSupport::TempCAGuard::~TempCAGuard() {
+    if (m_fd >= 0) {
+        unlink(m_fname.c_str());
+        close(m_fd);
+    }
+}
+
+
+XrdTpcNSSSupport::TempCAGuard::TempCAGuard(int fd, const std::string &fname)
+    : m_fd(fd), m_fname(fname)
+    {}
+
+
+XrdTpcNSSSupport::XrdTpcNSSSupport(XrdSysError *err, std::string ca_dir)
+    : m_log(*err),
+      m_ca_dir(ca_dir),
+      m_plugin(err, &g_ver, "Crypto Support", "libXrdCryptossl.so")
+{
+    if (!(m_x509_to_file_func = m_plugin.Resolve("XrdCryptosslX509ToFile")) ||
+        !(m_file_to_x509_func = m_plugin.Resolve("XrdCryptosslX509ParseFile")))
+    {
+        return;
+    }
+
+    Maintenance();
+}
+
+
+bool
+XrdTpcNSSSupport::NeedsNSSHack()
+{
+    const auto *version_info = curl_version_info(CURLVERSION_NOW);
+
+    if (!version_info->ssl_version) {return true;}
+    std::string ssl_version(version_info->ssl_version);
+
+    auto iter = ssl_version.find("NSS");
+    return iter != std::string::npos;
+}
+
+
+bool
+XrdTpcNSSSupport::Maintenance()
```

Comment:
I know you are going to hate it but there is a large potential of leaking file descriptors all over this code. Please use the functions in XrdSysFD.hh which makes sure that file descriptors are not passed to forked processes.

Reply:
Sure, good idea.  Will work on that.

Reply:
Done.  In the end, there were only two points where file descriptors were opened (the rest were just converting to things like `FILE*` and `DIR*`.

Necessitated the addition of a new `XrdSysFD_Openat` function; from a quick check of manpages, appears to be portable across the BSDs and Solaris so I kept it out of the Linux-specific block.